{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import time\n",
    "import torch\n",
    "import argparse\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from torch import Tensor, nn\n",
    "from datetime import datetime\n",
    "import matplotlib.pyplot as plt\n",
    "from torchinfo import summary  # 需要安装 torchinfo 包\n",
    "from torch.utils.data import random_split, DataLoader, Dataset\n",
    "from scipy.stats import spearmanr\n",
    "import warnings\n",
    "\n",
    "import torch.optim\n",
    "\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"] = \"0\" \n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "import grelu.lightning\n",
    "import grelu.data\n",
    "import grelu.model.models\n",
    "from grelu.sequence.format import (\n",
    "    INDEX_TO_BASE_HASH,\n",
    "    indices_to_one_hot,\n",
    "    strings_to_indices,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GeneExpressionDataset_self(Dataset):\n",
    "    def __init__(self, sequences, labels):\n",
    "        self.sequences = sequences\n",
    "        self.labels = labels\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.sequences)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "\n",
    "        # One-hot encode\n",
    "        seq = self.sequences[idx]\n",
    "        seq_strings = strings_to_indices(seq)\n",
    "        seq_one_hot = indices_to_one_hot(seq_strings)\n",
    "        \n",
    "        label = self.labels[idx]\n",
    "\n",
    "        return seq_one_hot, label\n",
    "\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_type = \"power\"\n",
    "folder_path = f\"code/model/CNN_Model\"\n",
    "save_dir = \"2025_2_4_1_35_power\" \n",
    "model_file = \"checkpoint.ckpt\"    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "==========================================================================================\n",
       "Layer (type:depth-idx)                   Output Shape              Param #\n",
       "==========================================================================================\n",
       "DNA_Conv1D_Model                         [1, 78]                   --\n",
       "├─Conv1d: 1-1                            [1, 64, 196608]           3,904\n",
       "├─BatchNorm1d: 1-2                       [1, 64, 196608]           128\n",
       "├─MaxPool1d: 1-3                         [1, 64, 39321]            --\n",
       "├─Dropout: 1-4                           [1, 64, 39321]            --\n",
       "├─Conv1d: 1-5                            [1, 128, 39321]           123,008\n",
       "├─BatchNorm1d: 1-6                       [1, 128, 39321]           256\n",
       "├─MaxPool1d: 1-7                         [1, 128, 7864]            --\n",
       "├─Dropout: 1-8                           [1, 128, 7864]            --\n",
       "├─Conv1d: 1-9                            [1, 128, 7864]            245,888\n",
       "├─BatchNorm1d: 1-10                      [1, 128, 7864]            256\n",
       "├─MaxPool1d: 1-11                        [1, 128, 1572]            --\n",
       "├─Dropout: 1-12                          [1, 128, 1572]            --\n",
       "├─Conv1d: 1-13                           [1, 32, 1572]             61,472\n",
       "├─BatchNorm1d: 1-14                      [1, 32, 1572]             64\n",
       "├─MaxPool1d: 1-15                        [1, 32, 314]              --\n",
       "├─Dropout: 1-16                          [1, 32, 314]              --\n",
       "├─Linear: 1-17                           [1, 1024]                 10,290,176\n",
       "├─Linear: 1-18                           [1, 78]                   79,950\n",
       "==========================================================================================\n",
       "Total params: 10,805,102\n",
       "Trainable params: 10,805,102\n",
       "Non-trainable params: 0\n",
       "Total mult-adds (G): 7.65\n",
       "==========================================================================================\n",
       "Input size (MB): 3.15\n",
       "Forward/backward pass size (MB): 298.78\n",
       "Params size (MB): 43.22\n",
       "Estimated Total Size (MB): 345.14\n",
       "=========================================================================================="
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = grelu.model.models.DNA_Conv1D_Model(expression_type).to(device)\n",
    "\n",
    "from torchinfo import summary  \n",
    "summary(model, input_size=(1, 4, 196608))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "if expression_type == \"power\": \n",
    "    mmc2_gene_expression_file = \"data/mmc2_78_gene_expression_power_transformed.npy\"\n",
    "elif expression_type == \"identity\":\n",
    "    mmc2_gene_expression_file = \"data/mmc2_78_gene_expression_identity_transformed.npy\"\n",
    "\n",
    "mmc2_gene_sequence_file   = \"data/mmc2_78_gene_sequence.fasta\"\n",
    "mmc2_gene_location_file   = \"data/mmc2_78_gene_location.csv\"\n",
    "\n",
    "# select gene sequence in length less than 100000 \n",
    "threshold = 100000 \n",
    "df_mmc2_gene_location = pd.read_csv(mmc2_gene_location_file)\n",
    "cell_type_list = df_mmc2_gene_location.iloc[:, 0].values\n",
    "df_mmc2_gene_location['len'] = pd.to_numeric(df_mmc2_gene_location['len'], errors='coerce')  # 如果有非数值的值，设置为NaN\n",
    "df_mmc2_gene_location = df_mmc2_gene_location[df_mmc2_gene_location['len'] < threshold]\n",
    "selected_gene = df_mmc2_gene_location[\"gene_id\"].to_list()\n",
    "\n",
    "# load gene expression \n",
    "mmc2_gene_expression = np.load(mmc2_gene_expression_file, allow_pickle=True)\n",
    "mmc2_gene_expression_dict = dict(zip(mmc2_gene_expression.item()[\"genes\"], mmc2_gene_expression.item()[\"expression\"]))\n",
    "\n",
    "# load gene sequence \n",
    "mmc2_gene_sequence_dict = {}\n",
    "for record in SeqIO.parse(mmc2_gene_sequence_file, \"fasta\"):\n",
    "    gene_id = record.id  \n",
    "    gene_sequence = str(record.seq) \n",
    "    mmc2_gene_sequence_dict[gene_id] = gene_sequence\n",
    "\n",
    "dataset_dict = {\n",
    "                'gene': [],\n",
    "                'sequence': [],\n",
    "                'expression': []\n",
    "            }\n",
    "dataset_df = pd.DataFrame(dataset_dict)\n",
    "\n",
    "for gene in selected_gene:\n",
    "    sequence = mmc2_gene_sequence_dict[gene]\n",
    "    expression = mmc2_gene_expression_dict[gene]\n",
    "\n",
    "    dataset_dict['gene'].append(gene)\n",
    "    dataset_dict['sequence'].append(sequence)\n",
    "    dataset_dict['expression'].append(expression)\n",
    "\n",
    "# tarin_step\n",
    "total_dataset = GeneExpressionDataset_self(dataset_dict['sequence'], dataset_dict['expression'])\n",
    "total_loader = DataLoader(total_dataset, batch_size=32, shuffle=True)\n",
    "\n",
    "def calculate_spearman(real_activate, fake_activate):\n",
    "\n",
    "    spearman_coefficients = []\n",
    "    real_tensor = torch.stack(real_activate)  \n",
    "    fake_tensor = torch.stack(fake_activate)\n",
    "\n",
    "    for col in range(real_tensor.shape[1]):\n",
    "        real_col = real_tensor[:, col].numpy()  \n",
    "        fake_col = fake_tensor[:, col].numpy()  \n",
    "        coefficient, _ = spearmanr(real_col, fake_col)\n",
    "        spearman_coefficients.append(coefficient)\n",
    "\n",
    "    return spearman_coefficients\n",
    "\n",
    "\n",
    "def Measure_Metric(model, total_loader):\n",
    "    model.eval()  \n",
    "    real_activate = []\n",
    "    fake_activate = []\n",
    "    with torch.no_grad():  \n",
    "        for inputs, targets in total_loader:\n",
    "            \n",
    "            inputs = torch.tensor(inputs, dtype=torch.float32).to(device)\n",
    "            outputs = model(inputs)\n",
    "            real_activate.extend(targets)\n",
    "            fake_activate.extend(outputs.to('cpu'))\n",
    "\n",
    "    spearman_coefficients = calculate_spearman(real_activate, fake_activate)\n",
    "\n",
    "    return spearman_coefficients, real_activate, fake_activate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = grelu.model.models.DNA_Conv1D_Model(expression_type).to(device)\n",
    "\n",
    "checkpoint = torch.load(os.path.join(folder_path, save_dir, model_file))\n",
    "model.load_state_dict(checkpoint['model_state_dict'])\n",
    "\n",
    "spearmanr_coefficients, real_activate, fake_activate = Measure_Metric(model, total_loader)\n",
    "\n",
    "def plt_figure(save_figure_path, cell_type, real_col, fake_col, spearmanr):\n",
    "\n",
    "    os.makedirs(save_figure_path, exist_ok=True)\n",
    "\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    plt.scatter(real_col, fake_col, color=\"blue\", alpha=0.5)\n",
    "\n",
    "    plt.title(f\"{cell_type}_{spearmanr:.5f}\", fontsize=16)\n",
    "    plt.xlabel(\"real\", fontsize=14)\n",
    "    plt.ylabel(\"fake\", fontsize=14)\n",
    "    plt.xlim(0, 200)  \n",
    "    plt.ylim(0, 200)  \n",
    "\n",
    "    plt.savefig(f'{save_figure_path}/{cell_type}_{spearmanr:.5f}.png')\n",
    "    plt.close()\n",
    "\n",
    "\n",
    "real_tensor = torch.stack(real_activate)  \n",
    "fake_tensor = torch.stack(fake_activate)\n",
    "\n",
    "for index in range(real_tensor.shape[1]):\n",
    "    real_col = real_tensor[:, index].numpy()  \n",
    "    fake_col = fake_tensor[:, index].numpy() \n",
    "\n",
    "    spearmanr = spearmanr_coefficients[index]\n",
    "    cell_type = cell_type_list[index]\n",
    "    plt_figure(os.path.join(folder_path, save_dir, \"figure\"), cell_type, real_col, fake_col, spearmanr)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Test_Demo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
